### PR TITLE
Convert trait/missile facings and rotation speeds to WAngle

### LINKED ISSUE #15196: Polish turn speed values
So, from what I see on the traits page, the maximun turn speed value allowed is 255, but from the testing im doing, the value 40 already gives you an almost instant turn rate, this means 215 numbers are not used, meanwhile the difference between 1 and 2 is so big that I wish there was a number in between.

On this issue I am suggesting to change the turn speed equation a bit so it uses the whole range of numbers more consistently, why have 255 numbers if you are not going to go beyond 20, which yeah, covers the need for instant turn rates, but leaves very few options for slow turn rates.

And now I am going to post a few gifs to elaborate the idea (its around 2.2 mb)

Turn Speed 30
turn speed 30
Turn Speed 20
turn speed 20
Turn Speed 10
turn speed 10
Turn Speed 5
turn speed 5

Edit: The unit at the left has TurnSpeed 2, while the one at the right has TurnSpeed: 1
turnspeed1and2

### PR DESCRIPTION
**Pull request by pchote**
This PR takes one of the final steps to completing the facing &rarr; WAngle conversion. Now only turrets, missiles, and a couple of other small pieces are left using the old system.

Closes #15196.

I have converted everything that ran through `WAngle.FromFacing` - it is possible that I have missed one or two things that kept everything as integers, but all the previous PRs have hopefully already removed those cases.

### CODE REVIEW

File: OpenRA.Mods.Common/Traits/Air/Aircraft.cs [hunk 1]
```diff
@@ -47,12 +47,12 @@ public class AircraftInfo : PausableConditionalTraitInfo, IPositionableInfo, IFa
 		[Desc("The speed at which the aircraft is repulsed from other aircraft. Specify -1 for normal movement speed.")]
 		public readonly int RepulsionSpeed = -1;
 
-		public readonly int InitialFacing = 0;
+		public readonly WAngle InitialFacing = WAngle.Zero;
 
-		public readonly int TurnSpeed = 255;
+		public readonly WAngle TurnSpeed = new WAngle(1023);
 
 		[Desc("Turn speed to apply when aircraft flies in circles while idle. Defaults to TurnSpeed if negative.")]
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Air/Aircraft.cs]: "Defaults to TurnSpeed if undefined".

This is a trivial change, so will leave it until there are other review comments.

Comment (reaperrr) [inline OpenRA.Mods.Common/Traits/Air/Aircraft.cs]: Should be "Defaults to TurnSpeed if undefined." now, I guess.

Edit: OK, didn't see your comment.

**Comment by reaperrr** [comment]
Looks good to me, won't have time to test it before Sunday next week at the earliest, though. Feel free to count this as +1 if it needs to be merged earlier.

File: OpenRA.Mods.Common/Traits/Mobile.cs [hunk 2]
```diff
@@ -30,10 +30,10 @@ public class MobileInfo : PausableConditionalTraitInfo, IMoveInfo, IPositionable
 		[Desc("Which Locomotor does this trait use. Must be defined on the World actor.")]
 		public readonly string Locomotor = null;
 
-		public readonly int InitialFacing = 0;
+		public readonly WAngle InitialFacing = WAngle.Zero;
 
 		[Desc("Speed at which the actor turns.")]
-		public readonly int TurnSpeed = 255;
+		public readonly WAngle TurnSpeed = new WAngle(1023);
```
Comment (tovl) [inline OpenRA.Mods.Common/Traits/Mobile.cs]: Anything higher than 512 is pointless, so maybe we should change the default value here?

File: OpenRA.Mods.Common/Traits/Turreted.cs [hunk 3]
```diff
@@ -21,8 +21,8 @@ public class TurretedInfo : PausableConditionalTraitInfo, Requires<BodyOrientati
 	{
 		public readonly string Turret = "primary";
 		[Desc("Speed at which the turret turns.")]
-		public readonly int TurnSpeed = 255;
-		public readonly int InitialFacing = 0;
+		public readonly WAngle TurnSpeed = new WAngle(1023);
```
Comment (tovl) [inline OpenRA.Mods.Common/Traits/Turreted.cs]: Same here.

File: OpenRA.Mods.Common/Traits/Air/Aircraft.cs [hunk 4]
```diff
@@ -47,12 +47,12 @@ public class AircraftInfo : PausableConditionalTraitInfo, IPositionableInfo, IFa
 		[Desc("The speed at which the aircraft is repulsed from other aircraft. Specify -1 for normal movement speed.")]
 		public readonly int RepulsionSpeed = -1;
 
-		public readonly int InitialFacing = 0;
+		public readonly WAngle InitialFacing = WAngle.Zero;
 
-		public readonly int TurnSpeed = 255;
+		public readonly WAngle TurnSpeed = new WAngle(1023);
```
Comment (tovl) [inline OpenRA.Mods.Common/Traits/Air/Aircraft.cs]: and here

**Comment by tovl** [comment]
We should aim to get this merged before continuing with #18322

File: OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnDeploy.cs [hunk 5]
```diff
@@ -109,7 +109,7 @@ public GrantConditionOnDeploy(ActorInitializer init, GrantConditionOnDeployInfo
 		{
 			self = init.Self;
 			checkTerrainType = info.AllowedTerrainTypes.Count > 0;
-			canTurn = self.Info.HasTraitInfo<IFacingInfo>();
+			canTurn = info.Facing.HasValue && self.Info.HasTraitInfo<IFacingInfo>();
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnDeploy.cs]: `canTurn` is now no longer used.

**Comment by pchote** [comment]
Fixed and rebased.

**Comment by pchote** [comment]
I missed `AttackPopupTurreted.DefaultFacing`.

**Comment by pchote** [comment]
Rebased and fixed.

File: mods/cnc/rules/structures.yaml [hunk 6]
```diff
@@ -1005,7 +1005,7 @@ GTWR:
 		Range: 3c0
 	WithMuzzleOverlay:
 	Turreted:
-		TurnSpeed: 255
+		TurnSpeed: 1023
```
Comment (abcdefg30) [inline mods/cnc/rules/structures.yaml]: This probably wants to be `512` now, too?

**Comment by pchote** [comment]
Updated gwtr, atwr, msam turret speeds.

File: mods/ra/rules/structures.yaml [hunk 7]
```diff
@@ -859,7 +859,7 @@ FTUR:
 	WithBuildingBib:
 		HasMinibib: true
 	Turreted:
-		TurnSpeed: 255
+		TurnSpeed: 1023
```
Comment (abcdefg30) [inline mods/ra/rules/structures.yaml]: This here probably as well.

File: mods/ra/rules/structures.yaml [hunk 8]
```diff
@@ -761,7 +761,7 @@ HBOX:
 		Condition: cloak-force-disabled
 		ValidDamageStates: Critical
 	Turreted:
-		TurnSpeed: 255
+		TurnSpeed: 1023
```
Comment (abcdefg30) [inline mods/ra/rules/structures.yaml]: And here and above.

File: OpenRA.Mods.Common/Activities/Air/FallToEarth.cs [hunk 9]
```diff
@@ -49,14 +49,13 @@ public override bool Tick(Actor self)
 				return true;
 			}
 
-			if (info.MaximumSpinSpeed != 0)
+			if (!info.MaximumSpinSpeed.HasValue || info.MaximumSpinSpeed.Value != WAngle.Zero)
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Activities/Air/FallToEarth.cs]: Imho it would be better to write this as `if (acceleration != 0)`.

**Comment by pchote** [comment]
Fixed and rebased on top of the TD balance changes.

**Review by abcdefg30 — CHANGES_REQUESTED** [review]
The voxel harvester docking looks a bit off:
grafik

**Comment by pchote** [comment]
Forgot to account for `TiberianSunRefinery` which inherited `DockAngle` from `Refinery`. Fixed.

**Comment by pchote** [comment]
Rebased.

**Review by abcdefg30 — APPROVED** [review]

**Review by reaperrr — APPROVED** [review]

**Comment by reaperrr** [comment]
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/77f3eb5d61771b06792c3e6b3b025a4856156ca4...6187f63c7eeee1b13f1d042ac4be6d63c28a1a18

**Pull request merged**